beard/clustering/wrappers.py

```diff
@@ -121,25 +136,31 @@ def fit(self, X, y=None):
         :returns: self
         """
         X = np.array(X)
-        size = X.shape[0]
+        n_samples = X.shape[0]
+        Xr = X  # Xr as the original raw data
 
         # Build linkage matrix
         if self.affinity == "precomputed":
-            Xs = X
+            Xa = X  # Xa as the affinity matrix
             i, j = np.triu_indices(X.shape[0], k=1)
             X = X[i, j]
             self.linkage_ = hac.linkage(X, method=self.method)
 
         elif callable(self.affinity):
             X = self.affinity(X)
-            Xs = X
+            Xa = X
             i, j = np.triu_indices(X.shape[0], k=1)
             X = X[i, j]
             self.linkage_ = hac.linkage(X, method=self.method)
-        else:
+
+        elif self.scoring_data != "affinity":
             self.linkage_ = hac.linkage(X,
                                         method=self.method,
                                         metric=self.affinity)
+        else:
+            raise ValueError("Inputs combination error: Make sure to use\
+                a callable function for the affinity parameter. If you don't,\
+                you would need to set the scoring_data parameter to affinity.")
 
```

Comment:
I find the last `elif` statement difficult to understand. I would instead structure the code like this:

```
        Xr = X  # Xr as the original raw data
        Xa = None

        # Build linkage matrix
        if self.affinity == "precomputed":
            Xa = X  # Xa as the affinity matrix
            i, j = np.triu_indices(X.shape[0], k=1)
            X = X[i, j]
            self.linkage_ = hac.linkage(X, method=self.method)

        elif callable(self.affinity):
            X = self.affinity(X)
            Xa = X
            i, j = np.triu_indices(X.shape[0], k=1)
            X = X[i, j]
            self.linkage_ = hac.linkage(X, method=self.method)

        else:
            self.linkage_ = hac.linkage(X,
                                        method=self.method,
                                        metric=self.affinity)
```

and then below, when you call the scoring function, check if `scoring_data == "affinity" and Xa is None`, if which case, trigger the exception.